1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+---
+RFC: RFC00XX
+Author: Dongbo Wang
+Status: Draft
+Area: Microsoft.PowerShell.Core
+Comments Due: 9/22/2021
+Plan to implement: Yes
+---
+
+# Loading Modules into Isolated AssemblyLoadContext
+
+Assembly isolation is a missing feature in PowerShell.
+Today, a PowerShell process loads all assemblies in the default load context.
+When an assembly is already loaded, it's not allowed to load the same assembly again with a different version.
+This results in the PowerShell issue [#2083](https://github.com/PowerShell/PowerShell/issues/2083).
+Basically, when two modules depend on different versions of the same assembly,
+if the module that depends on the lower version is loaded first,
+the other module cannot be loaded in the same session anymore.
+
+The proposal is to allow Loading a module into a custom `AssemblyLoadContext`.
+All the module required assemblies, all the assemblies loaded by `Add-Type` in the module script,
+as well as assemblies produced by the PowerShell classes declared in the module,
+will be loaded into the custom load context.
+
+## Motivation
+
+    As a PowerShell user, I'm able to use modules that depend on different versions of the same assembly in the same session.
+
+## Specification
+
+1. Create the type `PSModuleLoadContext` that derives from `AssemblyLoadContext` (ALC).
+   This type will be the custom ALC used for the isolated modules.
+
+1. Update `Import-Module` with two new parameters:
+
+    ```powershell
+    ## Import the module to an isolated AssemblyLoadContext.
+    Import-Module [-Isolated]
+
+    ## Import the module to the specified AssemblyLoadContext.
+    Import-Module [[-AssemblyLoadContext] <PSModuleLoadContext>]
+    ```
+    - When `-Isolated` is specified,
+      PowerShell creates a custom `AssemblyLoadContext` instance of the type `PSModuleLoadContext`,
+      which will be associated with the module.
+
+    - When `-AssemblyLoadContext` is specified,
+      the given `PSModuleLoadContext` instance will be associated with the module.
+
+    Unless otherwise instructed by the module author,
+    all assembly loading triggered by loading the module or executing the module
+    will have the assembly loaded into the associated assembly load context.
+
+    > NOTE: A module author always can explicitly choose where to load an assembly, by:
+    > - calling `LoadXXX` methods directly on an `AssemblyLoadContext` instance;
+    > - changing the active assembly load context with `AssemblyLoadContext.EnterContextualReflection`;
+    > - calling `Assembly.LoadFrom` to load an assembly to the default context;
+    > - calling `Assembly.LoadFile` to load an assembly to a separate anonymous load context;
```

Comment:
Are we adding a parameter to `Add-Type` or updating `RequiredAssemblies` in a module manifest to select the ALC loading behaviour? Or are we forcing module authors to reach directly into .NET to define the load context behaviour?

I think it's worth updating module manifests to allow module _authors_ to define the default load context behaviour for themselves.

Reply:
The list of ways to explicitly load assemblies into a different assembly load context is just to call out that a module author can choose to not respect the module isolation enforced by PowerShell.

But I think in general, a module author doesn't need to care about assembly load context and should just let PowerShell handle that for it. (there will be exceptions for sure, such as the Az modules, they are planning a change to handle assembly conflicts in a finer granularity)

Reply:
> I think it's worth updating module manifests to allow module authors to define the default load context behavior for themselves.

Any reason why you think a module author should define the assembly loading behavior?
In my opinion, a module author doesn't know the environment where the module will be used in, and thus won't be able to know what would be the right assembly loading behavior in that environment. It should be the user to decide whether the module should be loaded in a separate load context.

Reply:
If a module author creates a module that they know will have issues (e.g., it ships a library that's already included in PowerShell, and requires a higher version than the one PowerShell comes with), then placing the burden on end-users of modules to know that they'll always need to explicitly import the module doesn't make much sense to me.

Also, consider a scenario where module A depends on modules B and C, and the author of module A knows there is an incompatibility between B and C; how would they be able to use this? I think supporting it in module manifests makes a lot more sense than just allowing end users to do it. 

ALCs are a pretty advanced concept in general. No matter how friendly we make it, I think it's highly likely that module authors will be more familiar with the concept than most end-users, and we should allow for them to make sensible decisions when authoring that can be declared in the module manifest; otherwise, dependencies would end up being 'hidden' in the psm1 since they can only handle it from script, and can't rely on autoloading anymore.

Without _some_ measure of support for customizing how autoloading treats certain modules, I think we're far more likely to cause issues for users.

Reply:
> If a module author creates a module that they know will have issues (e.g., it ships a library that's already included in PowerShell, and requires a higher version than the one PowerShell comes with)

This is a scenario where allowing the module author to decide the loading behavior makes some sense. However, the module may not have a version conflicting issue with a newer version of PowerShell, but it will still be imported to a separate ALC because it's configured so in the manifest.

This may not seem to be a big deal, but I would argue that many more module authors would choose to always have their module to be loaded into a separate ALC "just to be safe". Is that what we really want? IMHO, we should not encourage loading modules into separate ALC unless it's really needed, so as to avoid type identity issues as much as possible.

> Also, consider a scenario where module A depends on modules B and C, and the author of module A knows there is an incompatibility between B and C; how would they be able to use this?

Should we really enable user to do this? It seems wrong from the beginning to depend on modules that are not compatible. I think the goal is not to enable arbitrary use of incompatible assemblies/modules in PowerShell. Instead, we should target the common scenarios where a user would likely run into with the "conflicting assembly" issue.

-------------------------

But, on the other hand, If we do allow configurations in manifest, what configurations should be provided?

I think we should only allow a module author to decide if the module should be loaded in a isolated ALC and nothing more. I don't think finer granularity controls would worth the complexity.

Reply:
Dependency management in PowerShell is already pretty complicated, and I can see this being useful for module authors if we allow the module author to specify a default behaviour for their module.

I would think that (ideally) we'd add a new key to the module manifest, e.g. `LoadIsolated`, which specifies whether it simply loads in the global context (defaults to `$false`) as usual, or whether it loads in a new context (set to `$true` to enable).

Perhaps a similarly named key could be allowed for each entry of `RequiredModules` (etc). I think it makes sense for most NestedModules to share the module context, but I'm unsure if it makes sense to create "special" behaviour for some kinds of dependencies that aren't permissible for others; it might be more sensible to just allow it for any referenced modules and/or assemblies.

The issue I have with _not_ providing for it in a module manifest is that a module author may be forced to make a binary module load in a psm1 which it wouldn't otherwise really need, save for requiring the isolation context. It would not be a great UX for users if some modules have to document "auto-loading is broken because of X, Y, Z -- you need to manually use `Import-Module` with the correct flags to get this module to load" when we could provide the framework for them to give a better UX to their end users.